packages/x-charts/src/ChartsXAxis/ChartsXAxis.tsx

```diff
@@ -32,43 +31,76 @@ const useUtilityClasses = (ownerState: ChartsXAxisProps & { theme: Theme }) => {
   return composeClasses(slots, getAxisUtilityClass, classes);
 };
 
-type LabelExtraData = { width: number; height: number; skipLabel?: boolean };
+function useVisibleLabels({
+  xTicks,
+  textMapRef,
+  tickLabelStyle,
+  tickLabelInterval,
+  reverse,
+}: {
+  xTicks: TickItemType[];
+  textMapRef: React.RefObject<Map<number, SVGTextElement>>;
+  tickLabelStyle?: ChartsTextStyle;
+  tickLabelInterval: ChartsXAxisProps['tickLabelInterval'];
+  reverse?: boolean;
+}) {
+  const isLabelVisibilityUserDefined = typeof tickLabelInterval === 'function';
+  const textMeasurementCache = React.useRef<Map<number, { width: number; height: number }>>(
+    new Map(),
+  );
+  const mounted = useMounted(true);
+
+  // Note: by moving this into a hook the compiler doesn't notice the issue, but we shouldn't be able to access this in
+  // a ref.
+  const measure = React.useCallback(
+    (index: number) => {
+      const value =
+        textMeasurementCache.current.get(index) ?? textMapRef.current.get(index)?.getBBox();
```

Comment:
This breaks the rules of the compiler and I'm not sure how to fix it. 

`measure` is called during render, which means we're indirectly accessing a ref during render.